src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -191,61 +198,88 @@ protected Solution SetCurrentSolution(Solution solution)
-                },
-                (@this: this, transformation, kind, projectId, documentId));
+                });
         }
 
         /// <summary>
-        /// Applies specified transformation to <see cref="CurrentSolution"/>, updates <see cref="CurrentSolution"/> to the new value and raises a workspace change event of the specified kind.
+        /// Applies specified transformation to <see cref="CurrentSolution"/>, updates <see cref="CurrentSolution"/> to
+        /// the new value and performs a requested callback immediately before and after that update.  The callbacks
+        /// will be invoked atomically while while <see cref="_serializationLock"/> is being held.
         /// </summary>
         /// <param name="transformation">Solution transformation.</param>
+        /// <param name="onBeforeUpdate">Action to perform immediately prior to updating <see cref="CurrentSolution"/>.
+        /// The action will be passed the old <see cref="CurrentSolution"/> that will be replaced and the exact solution
+        /// it will be replaced with. The latter may be different than the solution returned by <paramref
+        /// name="transformation"/> as it will have its <see cref="Solution.WorkspaceVersion"/> updated
+        /// accordingly.</param>
         /// <param name="onAfterUpdate">Action to perform once <see cref="CurrentSolution"/> has been updated.  The
         /// action will be passed the old <see cref="CurrentSolution"/> that was just replaced and the exact solution it
         /// was replaced with. The latter may be different than the solution returned by <paramref
-        /// name="transformation"/> as it will have its <see cref="Solution.WorkspaceVersion"/> updated accordingly.
-        /// Updating the solution and invoking <paramref name="onAfterUpdate"/> will happen atomically while <see
-        /// cref="_serializationLock"/> is being held.</param>
-        internal bool SetCurrentSolution<TData>(
+        /// name="transformation"/> as it will have its <see cref="Solution.WorkspaceVersion"/> updated
+        /// accordingly.</param>
+        private protected bool SetCurrentSolution<TData>(
             Func<Solution, TData, Solution> transformation,
-            Action<Solution, Solution, TData> onAfterUpdate,
-            TData data)
+            TData data,
+            Action<Solution, Solution, TData>? onBeforeUpdate = null,
+            Action<Solution, Solution, TData>? onAfterUpdate = null)
         {
             Contract.ThrowIfNull(transformation);
 
-            var currentSolution = Volatile.Read(ref _latestSolution);
+            var oldSolution = Volatile.Read(ref _latestSolution);
+
+            // Ensure our event handlers are realized prior to taking this lock.  We don't want to deadlock trying
+            // to obtain them when calling one of our callbacks. See https://github.com/dotnet/roslyn/issues/64681
+            EnsureEventListeners();
```

Comment:
i'm hoping this fixes the deadlock issue.  we still have a reentrancy issue.  namely that the first time this is called, we will get the evnet handlers, which will call all the way back around again into UpdateCurrentSolutionOnOptionsChanged.  this will be able to proceed and issue the change to the workspace as we aren't holding the lock then.